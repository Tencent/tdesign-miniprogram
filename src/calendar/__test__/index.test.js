import simulate from 'miniprogram-simulate';
import path from 'path';

describe('calendar', () => {
  const calendar = load(path.resolve(__dirname, `../calendar`), 't-calendar');

  it(':base', () => {
    const id = simulate.load({
      template: `<t-calendar value="{{value}}" minDate="{{minDate}}" maxDate="{{maxDate}}" visible></t-calendar>`,
      data: {
        minDate: +new Date(2022, 9, 1),
        maxDate: +new Date(2022, 9, 31),
        value: +new Date(2022, 9, 16),
      },
      usingComponents: {
        't-calendar': calendar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(':action', async () => {
    const date = new Date(2022, 0, 2);
    const onConfirm = jest.fn();
    const id = simulate.load({
      template: `<t-calendar minDate="{{min}}" bind:confirm="onConfirm" id="base" visible></t-calendar>`,
      data: {
        min: date.getTime(),
      },
      methods: {
        onConfirm,
      },
      usingComponents: {
        't-calendar': calendar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $base = comp.querySelector('#base');
    const [$disbledItem, $activeItem] = $base.querySelectorAll('.t-calendar__dates-item');
    const $confirmBtn = $base.querySelector('.t-calendar__confirm-btn');

    $disbledItem.dispatchEvent('tap');
    await simulate.sleep();

    expect($base.instance.base.value).toBe(null);

    $activeItem.dispatchEvent('tap');
    await simulate.sleep();

    expect($base.instance.base.value).toStrictEqual(date);

    $confirmBtn.dispatchEvent('tap');
    await simulate.sleep();

    expect(onConfirm).toHaveBeenCalled();
  });

  it(':confirm', async () => {
    const date = new Date(2022, 0, 1);
    const id = simulate.load({
      template: `<t-calendar minDate="{{min}}" id="base" visible></t-calendar>`,
      data: {
        min: date.getTime(),
      },
      usingComponents: {
        't-calendar': calendar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $base = comp.querySelector('#base');
    const $closeBtn = $base.querySelector('.t-calendar__close-btn');

    $closeBtn.dispatchEvent('tap');
    await simulate.sleep();

    expect($base.instance.visible).toBeFalsy();
  });

  it(':custom button', async () => {
    const id = simulate.load({
      template: `<t-calendar confirm-btn="{{confirmBtn}}" id="base" visible></t-calendar>`,
      data: {
        confirmBtn: { content: 'ok' },
      },
      usingComponents: {
        't-calendar': calendar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $base = comp.querySelector('#base');
    const $confirmBtn = $base.querySelector('.t-calendar__confirm-btn');

    $confirmBtn.dispatchEvent('tap');
    await simulate.sleep();

    expect($confirmBtn.dom.textContent).toBe('ok');
  });
});
