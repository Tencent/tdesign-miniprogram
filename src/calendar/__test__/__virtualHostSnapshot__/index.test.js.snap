// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calendar :base 1`] = `
<main>
  <wx-view
    ariaModal="{{true}}"
    ariaRole="dialog"
    class="t-popup t-popup--bottom t-fade-enter t-fade-enter-active class t-class"
    style="z-index:11500;"
    bind:transitionend="onTransitionEnd"
  >
    <wx-view
      class="t-popup__content t-class-content"
    >
      <wx-view
        class="t-popup__close"
        bind:tap="handleClose"
      />
      <wx-view
        class="t-calendar t-calendar--popup class t-class"
        style=""
      >
        <wx-view
          class="t-calendar__title"
          tabindex="0"
        >
          <wx-view
            class="t-calendar t-calendar--popup class t-class"
            style=""
          >
            <wx-view />
            <wx-view />
            <wx-view />
            <wx-view />
            <wx-view />
            <wx-view />
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月1日, 星期六"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 1,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            1
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月2日, 星期日"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 2,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            2
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月3日, 星期一"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 3,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            3
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月4日, 星期二"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 4,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            4
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月5日, 星期三"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 5,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            5
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月6日, 星期四"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 6,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            6
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月7日, 星期五"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 7,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            7
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月8日, 星期六"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 8,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            8
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月9日, 星期日"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 9,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            9
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月10日, 星期一"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 10,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            10
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月11日, 星期二"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 11,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            11
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月12日, 星期三"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 12,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            12
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月13日, 星期四"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 13,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            13
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月14日, 星期五"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 14,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            14
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月15日, 星期六"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 15,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            15
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="已选中, 10月16日, 星期日"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--selected"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 16,
                  "type": "selected",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            16
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月17日, 星期一"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 17,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            17
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月18日, 星期二"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 18,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            18
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月19日, 星期三"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 19,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            19
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月20日, 星期四"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 20,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            20
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月21日, 星期五"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 21,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            21
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月22日, 星期六"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 22,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            22
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月23日, 星期日"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 23,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            23
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月24日, 星期一"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 24,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            24
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月25日, 星期二"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 25,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            25
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月26日, 星期三"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 26,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            26
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月27日, 星期四"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 27,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            27
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月28日, 星期五"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 28,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            28
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月29日, 星期六"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 29,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            29
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月30日, 星期日"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 30,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            30
            
            </wx-view>
            <wx-view
              ariaDisabled="{{false}}"
              ariaLabel="10月31日, 星期一"
              ariaRole="button"
              class="t-calendar__dates-item  t-calendar__dates-item--"
              data-date="{{
                Object {
                  "date": Object {},
                  "day": 31,
                  "type": "",
                }
              }}"
              data-month="{{9}}"
              data-year="{{2022}}"
              bind:tap="handleSelect"
            >
              
            31
            
            </wx-view>
          </wx-view>
          <wx-view
            class="t-popup__close"
            bind:tap="handleClose"
          />
        </wx-view>
      </wx-view>
    </wx-view>
  </wx-view>
  <wx-view
    ariaLabel="关闭"
    ariaRole="button"
    class="t-overlay t-fade-enter t-fade-enter-active class"
    style="z-index:11000;"
    bind:tap="handleClick"
    catch:touchmove="noop"
    bind:transitionend="onTransitionEnd"
  />
</main>
`;
